Put sparse all reduce results to input tensors

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32226 Put sparse all reduce results to input tensors**

right now if users call torch.dist.all_reduce() on dense tensors, outputs are put in input tensors. but if users call torch.dist.all_reduce() on sparse tensors, outputs are neither returned explicitly to users nor are put in input tensors.

To make torch.dist.all_reduce() API have same behavior on both dense tensors and sparse tensors, this diff is made to make torch.dist.all_reduce() on sparse tensors to put output in input tensors as well. This is acheived by simply calling input_sparse.copy_(output_sparse), see PR https://github.com/pytorch/pytorch/pull/32197 that implemented copy_ for sparse tensors.

close #31413

Differential Revision: [D19192952](https://our.internmc.facebook.com/intern/diff/D19192952/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19192952/)!

zhaojuanmao committed:
Put sparse all reduce results to input tensors

right now if users call torch.dist.all_reduce() on dense tensors, outputs are put in input tensors. but if users call torch.dist.all_reduce() on sparse tensors, outputs are neither returned explicitly to users nor are put in input tensors.

To make torch.dist.all_reduce() API have same behavior on both dense tensors and sparse tensors, this diff is made to make torch.dist.all_reduce() on sparse tensors to put output in input tensors as well. This is acheived by simply calling input_sparse.copy_(output_sparse), see PR https://github.com/pytorch/pytorch/pull/32197 that implemented copy_ for sparse tensors.

close #31413

Differential Revision: [D19192952](https://our.internmc.facebook.com/intern/diff/D19192952/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19192952/)!

[ghstack-poisoned]

mrshenli commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -1199,6 +1200,7 @@ class AsyncSparseAllreduceCUDAWork : public AsyncSparseAllreduceWork {
     for (size_t i = 0; i < inputs.size(); i++) {
       stream_guard.reset_stream(streams[i]);
       outputs.push_back(output.to(inputs[i].device(), /*non_blocking=*/true));
+      inputs[i].copy_(output, /* non_blocking */ true);
```
how does non_blocking work for sparse tensors? For dense CPU tensors, we will need to use pinned memory to make it non_blocking, but the in-place copy for sparse tensors could change the number of elements, which might change the storage, which might prevent us from pinning the memory??

zhaojuanmao replied:
hmm, I'm not sure about pinning the memory underneath. 

But output.to(inputs[i].device(), /*non_blocking=*/true) is also kind of sparse tensor non_blocking copy?

For safety, I will change this line to "inputs[i].copy_(outputs[i], /* non_blocking */ true);"

mrshenli replied:
It's just me don't understand how non_blocking works for sparse tensors. If it is dense and CPU to CPU, IIUC, non_blocking is ignored. And `non_blocking` only works between CPU and CUDA.

Yep, for now, removing non_blocking should be OK.

zhaojuanmao replied:
I moved this copy into synchronize() call, please see updated PR. because we have to copy btw devices only after previous non blocking copy call is finished. 

regarding how non_blocking works for sparse tensors, I searched a little bit and did not find how it works with pinned memory, but from the #9005, looks like indices, values are copied separately and they have to be coalesced first

zhaojuanmao committed:
Update on "Put sparse all reduce results to input tensors"


right now if users call torch.dist.all_reduce() on dense tensors, outputs are put in input tensors. but if users call torch.dist.all_reduce() on sparse tensors, outputs are neither returned explicitly to users nor are put in input tensors.

To make torch.dist.all_reduce() API have same behavior on both dense tensors and sparse tensors, this diff is made to make torch.dist.all_reduce() on sparse tensors to put output in input tensors as well. This is acheived by simply calling input_sparse.copy_(output_sparse), see PR https://github.com/pytorch/pytorch/pull/32197 that implemented copy_ for sparse tensors.

close #31413

Differential Revision: [D19192952](https://our.internmc.facebook.com/intern/diff/D19192952/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19192952/)!

[ghstack-poisoned]

mrshenli commented:
The PR (#32197) linked in the description might be wrong, the current link points to a torch script related PR

zhaojuanmao commented:
> The PR (#32197) linked in the description might be wrong, the current link points to a torch script related PR

Right, somehow my PR messed up earlier, I updated commit log with correct link, basically one sparse tensor can call copy_ to copy from another sparse tensor as long as the two tensors have the same sizes().

zhaojuanmao commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -976,6 +976,7 @@ class AsyncSparseAllreduceWork : public ProcessGroupGloo::AsyncWork {
     // Copy back to input tensors.
     outputs.reserve(inputs.size());
     for (size_t i = 0; i < inputs.size(); i++) {
+      inputs[i].copy_(output);
```
I think so, for cuda work, we still need to copy back data to input tensor

zhaojuanmao commented:
This is the PR #9005 that implemented copy_ for sparse tensors.

zhaojuanmao committed:
Update on "Put sparse all reduce results to input tensors"


right now if users call torch.dist.all_reduce() on dense tensors, outputs are put in input tensors. but if users call torch.dist.all_reduce() on sparse tensors, outputs are neither returned explicitly to users nor are put in input tensors.

To make torch.dist.all_reduce() API have same behavior on both dense tensors and sparse tensors, this diff is made to make torch.dist.all_reduce() on sparse tensors to put output in input tensors as well. This is acheived by simply calling input_sparse.copy_(output_sparse), see PR https://github.com/pytorch/pytorch/pull/32197 that implemented copy_ for sparse tensors.

close #31413

Differential Revision: [D19192952](https://our.internmc.facebook.com/intern/diff/D19192952/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19192952/)!

[ghstack-poisoned]

zhaojuanmao committed:
Update on "Put sparse all reduce results to input tensors"


right now if users call torch.dist.all_reduce() on dense tensors, outputs are put in input tensors. but if users call torch.dist.all_reduce() on sparse tensors, outputs are neither returned explicitly to users nor are put in input tensors.

To make torch.dist.all_reduce() API have same behavior on both dense tensors and sparse tensors, this diff is made to make torch.dist.all_reduce() on sparse tensors to put output in input tensors as well. This is acheived by simply calling input_sparse.copy_(output_sparse), see PR https://github.com/pytorch/pytorch/pull/32197 that implemented copy_ for sparse tensors.

close #31413

Differential Revision: [D19192952](https://our.internmc.facebook.com/intern/diff/D19192952/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19192952/)!

[ghstack-poisoned]

zhaojuanmao commented:
The test failures are because newly added tests are not skipped properly in non GPU and non Gloo backend test environment, will fix it

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -908,7 +908,6 @@ def all_reduce(tensor,
     else:
         work.wait()
 
-
```
irrelevant change?

mrshenli commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -1210,6 +1211,12 @@ class AsyncSparseAllreduceCUDAWork : public AsyncSparseAllreduceWork {
+   // Copy outputs back to inputs after synchronization, so that users can 
+   // access all reduce results from input tensors
+   for (size_t i = 0; i < inputs.size(); i++) {
+      inputs[i].copy_(outputs[i]);
```
Let's add a TODO here, this can potentially be non_blocking if we do pinned_memory correctly.

zhaojuanmao replied:
it is device to device copy, I think we do not need non_blocking here?

mrshenli replied:
I see. Yep, D2D copy will be synced on current streams on both devices.

mrshenli commented on test/test_distributed.py:
```diff
@@ -1013,6 +1014,29 @@ def test_all_reduce_full_group_max(self):
+            num_inputs=1)
+        for (inputs, outputs) in tests:
+            tensors = [fn(input) for input in inputs]
+            dist.all_reduce(tensors[0], dist.ReduceOp.SUM, group_id)
```
I might miss sth here, why this is only testing `tensors[0]`? What about other tensors?

zhaojuanmao replied:
because dist_all_reduce() only accepts one tensor right now? you can see in the test I intentionally only tested for inputs with size 1.

mrshenli commented on test/common_distributed.py:
```diff
@@ -110,6 +111,49 @@ def get_timeout(test_id):
     return TIMEOUT_OVERRIDE.get(test_id.split('.')[-1], TIMEOUT_DEFAULT)
 
 
+def simple_sparse_reduce_tests(rank, world_size, num_inputs=1):
```
Didn't look into this method, this is just moving the function and there is no changes here, right?

zhaojuanmao replied:
yes, just moved the function

mrshenli reviewed (approved):
LGTM! Thanks for fixing. I have some minor comments inline.

Stamped to unblock. Please fix test skips before landing. Thanks!

zhaojuanmao committed:
Update on "Put sparse all reduce results to input tensors"


right now if users call torch.dist.all_reduce() on dense tensors, outputs are put in input tensors. but if users call torch.dist.all_reduce() on sparse tensors, outputs are neither returned explicitly to users nor are put in input tensors.

To make torch.dist.all_reduce() API have same behavior on both dense tensors and sparse tensors, this diff is made to make torch.dist.all_reduce() on sparse tensors to put output in input tensors as well. This is acheived by simply calling input_sparse.copy_(output_sparse), see PR https://github.com/pytorch/pytorch/pull/32197 that implemented copy_ for sparse tensors.

close #31413

Differential Revision: [D19192952](https://our.internmc.facebook.com/intern/diff/D19192952/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19192952/)!

[ghstack-poisoned]

zhaojuanmao committed:
Update on "Put sparse all reduce results to input tensors"


right now if users call torch.dist.all_reduce() on dense tensors, outputs are put in input tensors. but if users call torch.dist.all_reduce() on sparse tensors, outputs are neither returned explicitly to users nor are put in input tensors.

To make torch.dist.all_reduce() API have same behavior on both dense tensors and sparse tensors, this diff is made to make torch.dist.all_reduce() on sparse tensors to put output in input tensors as well. This is acheived by simply calling input_sparse.copy_(output_sparse), see PR https://github.com/pytorch/pytorch/pull/32197 that implemented copy_ for sparse tensors.

close #31413

Differential Revision: [D19192952](https://our.internmc.facebook.com/intern/diff/D19192952/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19192952/)!

[ghstack-poisoned]
